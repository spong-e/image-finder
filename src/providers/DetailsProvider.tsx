import * as React from "react";
import { createContext, Dispatch, ReactNode, useReducer } from "react";
import detailsReducer from "./DetailsReducer";

const DetailsContext = createContext<DetailsState | undefined>(undefined);

const DetailsDispatchContext = createContext<
  Dispatch<DetailsAction> | undefined
>(undefined);

interface DetailsProviderProps {
  children: ReactNode;
}

function DetailsProvider({ children }: DetailsProviderProps) {
  const initialState: DetailsState = {
    details: {
      firstName: "",
      surname: "",
      topic: "",
      thumbnail: "",
    },
  };

  const [state, dispatch] = useReducer(detailsReducer, initialState);

  // NOTE: you *might* need to memoize this value
  // Learn more in http://kcd.im/optimize-context

  const value = { state, dispatch };
  return (
    <DetailsContext.Provider value={state}>
      <DetailsDispatchContext.Provider value={dispatch}>
        <>
          {JSON.stringify(state, null, 2)}
          {children}
        </>
      </DetailsDispatchContext.Provider>
    </DetailsContext.Provider>
  );
}

export { DetailsProvider, DetailsContext, DetailsDispatchContext };
