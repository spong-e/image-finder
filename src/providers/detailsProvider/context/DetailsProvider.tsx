import { createContext, Dispatch, FunctionComponent, useReducer } from "react";

import detailsReducer from "./DetailsReducer";

const DetailsContext = createContext<DetailsState | undefined>({
  details: {
    firstName: "",
    lastName: "",
    topic: "",
    thumbnail: "",
  },
});
const DetailsDispatchContext = createContext<
  Dispatch<DetailsAction> | undefined
>(undefined);

const DetailsProvider: FunctionComponent<DetailsProviderProps> = ({
  children,
}) => {
  const initialState: DetailsState = {
    details: {
      firstName: "",
      lastName: "",
      topic: "",
      thumbnail: "",
    },
  };

  const [state, dispatch] = useReducer(detailsReducer, initialState);

  return (
    <DetailsContext.Provider value={state}>
      <DetailsDispatchContext.Provider value={dispatch}>
        {children}
      </DetailsDispatchContext.Provider>
    </DetailsContext.Provider>
  );
};

export { DetailsContext, DetailsDispatchContext, DetailsProvider };
