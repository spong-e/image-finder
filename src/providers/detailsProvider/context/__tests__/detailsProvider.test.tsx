import { render } from "@testing-library/react";

import { DetailsProvider } from "../DetailsProvider";

describe("Details provider", () => {
  it("Provider renders children", () => {
    const { container } = render(
      <DetailsProvider>
        <h1>This is the Details Provider</h1>
      </DetailsProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
