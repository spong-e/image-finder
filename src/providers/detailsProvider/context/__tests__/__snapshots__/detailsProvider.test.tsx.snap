// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details provider Provider renders children 1`] = `
<div>
  <h1>
    This is the Details Provider
  </h1>
</div>
`;
